File: php/allergies/Allergies.php
<?php

class Allergen
{
    public const EGGS = 1;
    public const PEANUTS = 2;
    public const SHELLFISH = 4;
    public const STRAWBERRIES = 8;
    public const TOMATOES = 16;
    public const CHOCOLATE = 32;
    public const POLLEN = 64;
    public const CATS = 128;

    public function __construct(private int $score)
    {
    }

    public static function allergenList(): array
    {
        return array_map(static function ($allergen) {
            return new Allergen($allergen);
        }, (new ReflectionClass(__CLASS__))->getConstants());
    }

    public function getScore(): int
    {
        return $this->score;
    }
}

class Allergies
{
    public function __construct(private int $input)
    {
    }

    public function isAllergicTo(Allergen $allergen): bool
    {
        return $this->input & $allergen->getScore();
    }

    public function getList(): array
    {
        return array_filter(Allergen::allergenList(), function ($allergen) {
            return $this->isAllergicTo($allergen);
        });
    }
}

File: nim/allergies/allergies.nim
import std/bitops

type
  Allergen* = enum
    Eggs, Peanuts, Shellfish, Strawberries, Tomatoes, Chocolate, Pollen, Cats

proc isAllergicTo*(score: int, allergen: Allergen): bool =
  case allergen:
  of Eggs:
    return score.testBit(0)
  of Peanuts:
    return score.testBit(1)
  of Shellfish:
    return score.testBit(2)
  of Strawberries:
    return score.testBit(3)
  of Tomatoes:
    return score.testBit(4)
  of Chocolate:
    return score.testBit(5)
  of Pollen:
    return score.testBit(6)
  of Cats:
    return score.testBit(7)

proc allergies*(score: int): set[Allergen] =
  var ret: set[Allergen] = {}
  for a in Allergen:
    if isAllergicTo(score, a):
      ret = ret + {a}
  return ret
